### PR TITLE
Update lead status from salesforce

### PR DESCRIPTION
**Pull request by mwvolo**
A script to sync leads with Salesforce (to mark rejected leads as rejected in Accounts). I do not want to add this to cron yet. Will do that later after further checking that it's working as expected.

Refactored some of the lead creation code, it now creates a security log and sentry error when lead creation fails (was not doing this before, even though it looked like it was).

Also includes a task to create leads for users that should have got them but didn't. 

All tested on dev and looking good.

### CODE REVIEW

**Comment by mwvolo** [comment]
Should be the final PR, Dante. 🤞🏼  Thanks for all your time reviewing these.

File: app/routines/update_user_lead_info.rb [hunk 1]
```diff
@@ -0,0 +1,70 @@
+class UpdateUserLeadInfo
+  BATCH_SIZE = 250
+
+  def self.call
+    new.call
+  end
+
+  def call
+    loop do
+      users = User.where(salesforce_contact_id: nil).where.not(salesforce_lead_id: nil, role: :student, faculty_status: :rejected_faculty).limit(BATCH_SIZE)
```
[the remark this answers is not in the document]
Comment (mwvolo) [reply app/routines/update_user_lead_info.rb:10]: That seems like I would have to either load all the leads into memory, which will likely blow up, or make an API call for each user in the `find_each` loop.
Maybe I will modify this to work more like the contact sync, where it looks at the last modified date to load in a set of leads to be updated.

File: lib/tasks/accounts/create_leads_for_instructors_not_sent_to_sf.rake [hunk 2]
```diff
@@ -1,12 +1,30 @@
...
   # rake accounts:create_leads_for_instructors_not_sent_to_sf
-  # To be run to fix users that were not sent to SF during bug from 26Jan22 to xxFeb22 (hotfix release date)
   task :create_leads_for_instructors_not_sent_to_sf, [:day] => [:environment] do |t, args|
-    # get all the instructors that don't have lead or contact ids
-    users = User.where(salesforce_contact_id: nil, salesforce_lead_id: nil, role: :instructor, state: :activated).or(User.where.not(sheerid_verification_id: nil))
-    users.each { |user|
-      Newflow::CreateSalesforceLead.perform_later(user: user)
-    }
+    # get all the instructors that don't have lead or contact ids and have a complete profile
+    users = User.where(salesforce_contact_id: nil, salesforce_lead_id: nil, role: :instructor, is_profile_complete: true)
```
[the remark this answers is not in the document]
Comment (mwvolo) [reply lib/tasks/accounts/create_leads_for_instructors_not_sent_to_sf.rake:6]: I've checked production, there are only 967 that will get processed by this. 

I think we want the rejected in this case; although they were rejected, they should still have a matching lead in Salesforce. (though in checking, that only brings the total to 951, so only a few match this criteria).

File: app/routines/update_user_lead_info.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (Dantemss) [reply app/routines/update_user_lead_info.rb:10]: `find_in_batches` is also an option, it'll give you arrays of 1000 users (or some size you specify) so you can do 1 leads API call for each batch

File: app/routines/update_user_lead_info.rb [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  BATCH_SIZE = 250
+
+  def self.call
+    new.call
+  end
+  def call
+    users = User.where(salesforce_contact_id: nil)
+                .where.not(salesforce_lead_id: nil, role: :student, faculty_status: :rejected_faculty)
+
+    users.find_in_batches(batch_size: BATCH_SIZE) do |users|
```
Comment (mwvolo) [inline app/routines/update_user_lead_info.rb:11]: Does this look like it'll work as intended?

File: app/routines/update_user_lead_info.rb [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+                                              .where(accounts_uuid: users.map(&:uuid))
+                                              .to_a
+                                              .index_by(&:accounts_uuid)
+
+    users.map do |user|
+      lead = leads[user.uuid]
+
+      unless lead.nil?
+        previous_lead_id = user.salesforce_lead_id
+        user.salesforce_lead_id = lead.id # it might change in SF lead merging
```
Comment (Dantemss) [inline app/routines/update_user_lead_info.rb:29]: Doesn't matter too much but this line could be inside the `if` statement below

**Review by Dantemss — APPROVED** [review]

**Pull request merged**